--- fresh-tomato-clean/release/src-rt-6.x.4708/router/shared/id.c	2025-01-16 13:57:56.066265387 +1000
+++ fresh-tomato-modified/release/src-rt-6.x.4708/router/shared/id.c	2025-01-18 05:49:31.742215434 +1000
@@ -66,8 +66,9 @@
 R6700v3				BCM4708C0             0x0646       32        0x1601 // same as R6400v2
 XR300				BCM4708C0             0x0646       32        0x1601 // same as R6400v2
 
-DIR-868L			BCM4708               0x0646       24        0x1110
+DIR-868L			BCM4708               0x0665       32        0x1111 // Changed to incorrect values to avoid conflict
 DIR-868LC1			BCM4708               0x0646       24        0x1101 //same as rev a/b but different boardrev
+DSL2890AL			BCM4708		      0x0646       24        0x1110 // Values are correct, equal to original
 WS880				BCM4708               0x0646       1234      0x1101
 R1D				BCM4709               0x0665       32        0x1301 //same as R7000
 F9K1113v2			BCM40781A0            0x0646       02        0x1100    0x00000110  1:devid=0x43A9
@@ -147,8 +148,9 @@
 		if ((nvram_match("boardrev", "0x1301")) && (nvram_match("boardnum", "32")) && (nvram_match("board_id", "U12H270T11_NETGEAR"))) return MODEL_R6900;
 		if ((nvram_match("boardrev", "0x1301")) && (nvram_match("boardnum", "32")) && (nvram_match("board_id", "U12H270T00_NETGEAR"))) return MODEL_R7000;
 		if ((nvram_match("boardrev", "0x1100")) && (nvram_match("boardnum", "679")) && (nvram_match("board_id", "U12H317T00_NETGEAR"))) return MODEL_EX7000;
-		if ((nvram_match("boardrev", "0x1110")) && (nvram_match("boardnum", "24"))) return MODEL_DIR868L;
+		if ((nvram_match("boardrev", "0x1111")) && (nvram_match("boardnum", "32"))) return MODEL_DIR868L; // Also changed to incorrect values to avoid conflict
 		if ((nvram_match("boardrev", "0x1101")) && (nvram_match("boardnum", "24"))) return MODEL_DIR868L;  /* rev c --> almost the same like rev a/b but different boardrev */
+		if ((nvram_match("boardrev", "0x1110")) && (nvram_match("boardnum", "24"))) return MODEL_DSL2890AL;
 		if ((nvram_match("boardrev", "0x1101")) && (nvram_match("boardnum", "1234"))) return MODEL_WS880;
 		if ((nvram_match("boardtype","0xE646")) && (nvram_match("boardnum", "20140309"))) return MODEL_EA6350v1;
 		if ((nvram_match("boardtype","0xE646")) && (nvram_match("boardnum", "20130125"))) return MODEL_EA6350v1; /* EA6200 --> same like EA6350v1, AC1200 class router */
